### PR TITLE
FlowMLA: zero-overhead DP MLA throughput boost via memory optimization

### PR DESCRIPTION
**Pull request by Siritao**
## Motivation

This PR is part of our upcoming work FlowMLA @antgroup. Targeting MLA in Data Parallel scenarios, we enable serving DeepSeek-like models on fewer devices through seamless memory optimization, while providing larger batching capacity for enhanced throughput.
Specifically, we partition the `o_proj` weights—the largest memory consumer in DP attention—across multiple DP ranks. We non-blockingly prefetch these weights via `all_gather` only prior to each layer's computation. These gathered weights physically reuse the same GPU memory buffer across operations. This reduces per-rank parameter redundancy to `1/dp_size`, freeing up more memory for kv cache, which in turn enables significant throughput gains. Furthermore, our optimization now makes serving FP8 DeepSeek-R1 models on 8×H20 GPU instances feasible, where previously impossible.

display_hidden

The figure above shows the serving profile for DeepSeek-R1 FP8 on 8×H20 GPUs at a single request, demonstrating that prefetch communication can be effectively hidden.

Throughput experiment: 32×H800 (4 nodes, NVLink), DeepSeek-R1 BF16, tp32 moe + dp32 Attention, input length 1024, output length 2048, maximize batch size

| Metric                   | w/o FlowMLA | w/ FlowMLA |
|--------------------------|-------------|------------|
| Throughput (tokens/sec)  | 280         | 400        |

Note that the current version only supports DP Attention, not DP+TP.

## Modifications

To achieve this, we implemented the `vTensor` library for weight management. We added two server arguments (`--vtensor-enable` and `--vtensor-port`) to enable our optimization, partially modified the implementations of `deepseek_v2.py`, `RowParallelLinear` class, `cuda_graph_runner.py`, and added support for `fp8_e4m3`.

To run our Opt:​​
- Install the `vTensor` module: `cd sglang/vtensor && python setup.py install`
- When serving, add `--vtensor-enable` to your startup command, optionally specifying a port with `--vtensor-port [PORT]`

## Checklist

- [x] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by ch-wan** [comment]
Is it applied to prefill or decode?

**Comment by Siritao** [comment]
> Is it applied to prefill or decode?

The optimization applies to both prefill and decoding stages. For decoding, we add cudagraph support.

**Comment by ch-wan** [comment]
@Siritao Thank you for your excellent contribution. I'd like to get more details about the settings. What is the batch size of your evaluation? I assume that when bs is small, communication cannot be fully hidden. What is the threshold?

**Comment by Siritao** [comment]
> @Siritao Thank you for your excellent contribution. I'd like to get more details about the settings. What is the batch size of your evaluation? I assume that when bs is small, communication cannot be fully hidden. What is the threshold?

Thanks for your attention. The figure was traced on ​8xH20 (DeepSeek-R1 FP8) with `batch_size=1`, demonstrating that communication overhead can be effectively hidden even under minimal batch conditions. However, FlowMLA’s performance does vary based on GPUs and model scales. To address this, we’ve added ​large-scale throughput experiments​ on ​32xH800 (DeepSeek-R1 BF16)​ in the PR submission. For this test, we used `tp32` moe && `dp32` Attention​ (o_proj sharded intra-node) && fixed input/output length​ (1k/2k tokens) && maximize batch size.
In this test scenario, enabling FlowMLA increased KV cache size from 6.22 GB to 16.80 GB per device (with `--mem-fraction-static=0.9`) while significantly boosting throughput, demonstrating the effectiveness of FlowMLA for memory-bound workloads.

File: vtensor/vtensor.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    CUresult result = (call);                                                  \
+    if (CUDA_SUCCESS != result) {                                              \
+      const char *errMsg;                                                      \
+      cuGetErrorString(result, &errMsg);                                       \
+      ASSERT(0, "Error when exec " #call " %s-%d code:%d err:%s",              \
+             __FUNCTION__, __LINE__, result, errMsg);                          \
+    }                                                                          \
+  }
+
+class PhyBlock {
```
Comment (yiakwy-xpu-ml-framework-team) [inline vtensor/vtensor.h:39]: Hi, this memory block is not classic `cached` based design (implemented memroy block, segment logics). 

Instead of extending pytroch cached memory allocator, I am concerned that these memory blocks may not be correctly recorded by Pytorch event profiler.

Last year we were discussing in Pytorch community how to extend pytorch cached allocator for blocked based GPU memroy allocator (in application of training/inference) :

> https://github.com/pytorch/pytorch/issues/115336

At that time, direclty inheritance is not possible, so I rewrited a cached pytorch memory allocator, but I believe now we can directly inherited pytorch cached allocator interface. 

This will bring good features : allocation can be correctly recorded by pytorch event profiler and mangement of block based GPU memory can be minimized in allocator logics.

**Comment by yiakwy-xpu-ml-framework-team** [comment]
As offline discussed, mapping/unmapping these kind memory is essential to applicaiton such as RL (cc @zhaochenyang20 ), I recall the ref vllm [PR#11743 support release of KV cache memory in sleep mode](https://github.com/vllm-project/vllm/pull/11743) last year. 

My suggestion is to make a CachedCuMemAllocator which inherit pytorch CachedMemoryPlugin by maintaining a private memory pool. 

SInce this function, is not exclusive to this PR (this PR does not offload memory to CPU), we can do it in SGLang-Kernel with SGlang version cuMemory Allocator , and I can add cached version in the proposal.

To use pytorch cached memory allocate machenism, simply export `alloc`, `dealloc` external C API interface, and bind the function in python side :

```
def get_pluggable_allocator(
    python_malloc_fn: Callable[[int],
                               int], python_free_func: Callable[[int, int],
                                                                None]
) -> torch.cuda.memory.CUDAPluggableAllocator:
    init_module(python_malloc_fn, python_free_func)
    new_alloc = torch.cuda.memory.CUDAPluggableAllocator(
        lib_name, 'my_malloc', 'my_free')
    return new_alloc
```

cc @zhyncs

**Comment by miter6** [comment]
Dose flowmla support for deepep?

**Comment by miter6** [comment]
another question.  
Dose flowmla support for PD serving?
